### PR TITLE
Feature/#330 zakano algorithm

### PR DESCRIPTION
**Pull request by oconn**
# PR for issue #330 .

## Ready for a PR?
 - [x] I have updated the CHANGELOG with an entry including a description, a link to my profile, and a link to the issue ticket. This change is filed under an Enhancement or Bug Fix, is shown under the relevant release tag name, and is included in my PR branch. 
 
## Implementation Notes:

Pulled in `serializable-fn` to source the zakano code as a string. 

## Breaking changes:

None

### CODE REVIEW

File: src/wombats/game/zakano_code.clj [hunk 1]
```diff
@@ -267,19 +282,47 @@
...
+           current-coords
+           global-coords
+           global-dimensions
+           global-arena]}
+   look-ahead-distance]
+  (let [orientation
+        (get-in self [:contents :orientation])
+
+        look-ahead-coords
+        (loop [coords []
```
Comment (emilyseibert) [inline src/wombats/game/zakano_code.clj]: separate function

File: src/wombats/game/zakano_code.clj [hunk 2]
```diff
@@ -296,56 +339,54 @@
...
-        should-shoot? (some #(contains? look-ahead-items %) ["steel-barrier"
-                                                             "wood-barrier"
-                                                             "wombat"
-                                                             "zakano"])
-        should-turn? (contains? look-ahead-items "poison")]
-
+(defsource pathfinding-action
+  [enriched-state]
+  (let [look-ahead-items (set (get-look-ahead-items enriched-state 3))
+        should-shoot? (some (fn [element]
```
Comment (emilyseibert) [inline src/wombats/game/zakano_code.clj]: no `?`

File: src/wombats/game/zakano_code.clj [hunk 3]
```diff
@@ -296,56 +339,54 @@
...
+                              (contains? look-ahead-items element))
+                            ["wood-barrier"
+                             "wombat"
+                             "zakano"])
+        should-turn? (some (fn [element]
```
Comment (emilyseibert) [inline src/wombats/game/zakano_code.clj]: no `?`

File: src/wombats/game/zakano_code.clj [hunk 4]
```diff
@@ -296,56 +339,54 @@
...
                          should-shoot? {:action :shoot}
                          should-turn? {:action :turn
                                        :metadata {:direction :right}}
                          :else {:action :move})]}))
 
-(defn clueless-action
+(defsource fire-action
+  [enriched-state]
+  (let [look-ahead-items (set (get-look-ahead-items enriched-state 3))
+        should-shoot? (some (fn [element]
```
Comment (emilyseibert) [inline src/wombats/game/zakano_code.clj]: no `?`

File: src/wombats/game/zakano_code.clj [hunk 5]
```diff
@@ -379,15 +425,19 @@
...
 
-(defn calculate-next-command
+(defsource calculate-next-command
   [enriched-state]
-  (first (filter #(not (nil? %))
-                 (map #(xform-command enriched-state (:name %) (:fn %))
-                      command-priority))))
-
-(defn calculate-optimal-command
+  (first (filter (fn [command] (not (nil? command)))
```
Comment (emilyseibert) [inline src/wombats/game/zakano_code.clj]: some?

Comment (oconn) [inline src/wombats/game/zakano_code.clj]: `some?` returns a boolean. Here I'm looking to get the first item in the command vector. So this got me looking into `some?` a little more. (in particular, `some?` vs `some`. `some?` will always return a boolean, `some` will return `true` or `nil`

**Comment by oconn** [comment]
@emilyseibert @dehli Added additional logging to capture runtime errors during Zakano code execution.

File: test/wombats/test/game/zakano_code.clj [hunk 6]
```diff
@@ -114,6 +106,11 @@
     (is (= :e (zc/modify-orientation :w :about-face)))
     (is (= :n (zc/modify-orientation :s :about-face)))))
 
+(deftest get-look-ahead-items
+  (testing "returns the next n items in front of the character"
+    (is (= ["open" "open" "open"]
+           (zc/get-look-ahead-items (zc/enrich-state sample-state)
```
Comment (dehli) [inline test/wombats/test/game/zakano_code.clj]: That's cool that Clojure does elementwise comparison for vecs.

**Pull request merged**